Sources/Basics/AuthorizationProvider.swift

```diff
@@ -1,19 +1,37 @@
 /*
  This source file is part of the Swift.org open source project
+
  Copyright (c) 2021 Apple Inc. and the Swift project authors
  Licensed under Apache License v2.0 with Runtime Library Exception
+
  See http://swift.org/LICENSE.txt for license information
  See http://swift.org/CONTRIBUTORS.txt for Swift project authors
  */
 
-import Foundation
+import struct Foundation.Data
+import struct Foundation.URL
+#if os(macOS)
+import Security
+#endif
+
+import TSCBasic
+import TSCUtility
 
 public protocol AuthorizationProvider {
-    func authentication(for url: URL) -> (user: String, password: String)?
+    mutating func addOrUpdate(for url: Foundation.URL, user: String, password: String, callback: @escaping (Result<Void, Error>) -> Void)
```

Comment:
what is the background to add the mutating aspect?

Reply:
I don't like having this `mutating` either. The reason for it is because `Netrc`'s state changes with `addOrUpdate`, and we need `addOrUpdate` to support registry auth:

> When credentials are provided, the corresponding object in the registries.json file includes a login key with the passed value. If the project's .netrc file has an existing entry for a given machine and login, it's updated with the new password; otherwise, a new entry is added. If no .netrc file exists, a new one is created and populated with the new entry.

Reply:
I assume this text is from SE-0292? which part of the code would be triggering the "write' in this case? I wonder maybe that code can interact with the Netrc utility directly to do write, and the provider protocol can stay readonly.

Reply:
I guess what I am trying to say is that the "write" flow is not entirely clear to me

Reply:
> I assume this text is from SE-0292? 

Yes

> which part of the code would be triggering the "write' in this case?

In [SwiftPackageRegistryTool](https://github.com/apple/swift-package-manager/blob/main/Sources/Commands/SwiftPackageRegistryTool.swift#L110) I would think, and potentially add keychain as an option. cc @mattt 

> I guess what I am trying to say is that the "write" flow is not entirely clear to me

Same here, but it's clear that we'd need to support "write".

Reply:
> > I guess what I am trying to say is that the "write" flow is not entirely clear to me
> Same here, but it's clear that we'd need to support "write".

I think for the sake of this PR, I would remove the write function from the provider Api, but keep the implementation in the respective implementations, and then we can discuss with @mattt what he had in mind from a flow point of view. wdyt?

Reply:
> I don't like having this `mutating` either. The reason for it is because `Netrc`'s state changes with `addOrUpdate`, and we need `addOrUpdate` to support registry auth:

We decided to punt on this functionality for now (see https://github.com/apple/swift-package-manager/pull/3647#issuecomment-895549931). And we probably need to rethink this anyway if we're supporting both keychain and `.netrc`, because it's unclear where those supplied credentials should be written.

I agree with @tomerd that we should remove the mutating function from the provider protocol, but keep the underlying implementation. That way, if / when it comes time to implement that flow, we'll have everything we need.

Reply:
OK. I've removed `addOrUpdate` from the protocol.